File: Sources/SwiftMTH/MSDPConverter.swift
// MSDP protocol control bytes
private let MSDP_VAR: UInt8 = 1
private let MSDP_VAL: UInt8 = 2
private let MSDP_TABLE_OPEN: UInt8 = 3
private let MSDP_TABLE_CLOSE: UInt8 = 4
private let MSDP_ARRAY_OPEN: UInt8 = 5
private let MSDP_ARRAY_CLOSE: UInt8 = 6

// Telnet framing bytes
private let IAC: UInt8 = 255
private let SB: UInt8 = 250
private let SE: UInt8 = 240
private let TELOPT_MSDP: UInt8 = 69
private let TELOPT_GMCP: UInt8 = 201

/// Convert MSDP binary subnegotiation to GMCP JSON subnegotiation.
///
/// Input: `IAC SB TELOPT_MSDP <msdp data> IAC SE`
/// Output: `IAC SB TELOPT_GMCP MSDP {<json>} IAC SE`
public func msdp2json(_ src: [UInt8]) -> [UInt8] {
    var out: [UInt8] = []
    out.reserveCapacity(src.count * 2)
    let srclen = src.count

    // If MSDP framing, replace with GMCP framing + "MSDP {"
    if srclen >= 3 && src[2] == TELOPT_MSDP {
        out.append(IAC)
        out.append(SB)
        out.append(TELOPT_GMCP)
        out.append(contentsOf: "MSDP {".utf8)
    }

    var i = 3
    var nest = 0
    var last: UInt8 = 0

    while i < srclen {
        if src[i] == IAC && i + 1 < srclen && src[i + 1] == SE {
            break
        }

        switch src[i] {
        case MSDP_TABLE_OPEN:
            out.append(UInt8(ascii: "{"))
            nest += 1
            last = MSDP_TABLE_OPEN

        case MSDP_TABLE_CLOSE:
            if last == MSDP_VAL || last == MSDP_VAR {
                out.append(UInt8(ascii: "\""))
            }
            if nest > 0 { nest -= 1 }
            out.append(UInt8(ascii: "}"))
            last = MSDP_TABLE_CLOSE

        case MSDP_ARRAY_OPEN:
            out.append(UInt8(ascii: "["))
            nest += 1
            last = MSDP_ARRAY_OPEN

        case MSDP_ARRAY_CLOSE:
            if last == MSDP_VAL || last == MSDP_VAR {
                out.append(UInt8(ascii: "\""))
            }
            if nest > 0 { nest -= 1 }
            out.append(UInt8(ascii: "]"))
            last = MSDP_ARRAY_CLOSE

        case MSDP_VAR:
            if last == MSDP_VAL || last == MSDP_VAR {
                out.append(UInt8(ascii: "\""))
            }
            if last == MSDP_VAL || last == MSDP_VAR || last == MSDP_TABLE_CLOSE || last == MSDP_ARRAY_CLOSE {
                out.append(UInt8(ascii: ","))
            }
            out.append(UInt8(ascii: "\""))
            last = MSDP_VAR

        case MSDP_VAL:
            if last == MSDP_VAR {
                out.append(UInt8(ascii: "\""))
                out.append(UInt8(ascii: ":"))
            }
            if last == MSDP_VAL {
                out.append(UInt8(ascii: "\""))
                out.append(UInt8(ascii: ","))
            }
            if i + 1 < srclen && src[i + 1] != MSDP_TABLE_OPEN && src[i + 1] != MSDP_ARRAY_OPEN {
                out.append(UInt8(ascii: "\""))
            }
            last = MSDP_VAL

        case UInt8(ascii: "\\"):
            out.append(UInt8(ascii: "\\"))
            out.append(UInt8(ascii: "\\"))

        case UInt8(ascii: "\""):
            out.append(UInt8(ascii: "\\"))
            out.append(UInt8(ascii: "\""))

        default:
            out.append(src[i])
        }
        i += 1
    }

    // Append closing "}" and IAC SE
    out.append(UInt8(ascii: "}"))
    out.append(IAC)
    out.append(SE)

    return out
}

/// Convert GMCP JSON subnegotiation to MSDP binary subnegotiation.
///
/// Input: `IAC SB TELOPT_GMCP MSDP {<json>} IAC SE`
/// Output: `IAC SB TELOPT_MSDP <msdp data> IAC SE`
public func json2msdp(_ src: [UInt8]) -> [UInt8] {
    var out: [UInt8] = []
    out.reserveCapacity(src.count)
    let srclen = src.count

    // If GMCP framing, replace with MSDP framing
    if srclen >= 3 && src[2] == TELOPT_GMCP {
        out.append(IAC)
        out.append(SB)
        out.append(TELOPT_MSDP)
    }

    var i = 3

    // Skip "MSDP {" prefix if present
    if i + 6 <= srclen {
        let prefix = src[i..<i+6]
        if prefix.elementsEqual("MSDP {".utf8) {
            i += 6
        }
    }

    var state = [Int](repeating: 0, count: 100)
    var nest = 0
    var last: UInt8 = 0
    state[0] = 0

    while i < srclen && src[i] != IAC && nest < 99 {
        switch src[i] {
        case UInt8(ascii: " "):
            i += 1

        case UInt8(ascii: "{"):
            out.append(MSDP_TABLE_OPEN)
            i += 1
            nest += 1
            state[nest] = 0

        case UInt8(ascii: "}"):
            nest -= 1
            i += 1
            if nest < 0 {
                out.append(IAC)
                out.append(SE)
                return out
            }
            out.append(MSDP_TABLE_CLOSE)

        case UInt8(ascii: "["):
            i += 1
            nest += 1
            state[nest] = 1
            out.append(MSDP_ARRAY_OPEN)

        case UInt8(ascii: "]"):
            nest -= 1
            i += 1
            out.append(MSDP_ARRAY_CLOSE)

        case UInt8(ascii: ":"):
            out.append(MSDP_VAL)
            i += 1

        case UInt8(ascii: ","):
            i += 1
            if state[nest] != 0 {
                out.append(MSDP_VAL)
            } else {
                out.append(MSDP_VAR)
            }

        case UInt8(ascii: "\""):
            i += 1
            if last == 0 {
                last = MSDP_VAR
                out.append(MSDP_VAR)
            }

            // Read quoted string
            var reading = true
            while i < srclen && src[i] != IAC && reading {
                switch src[i] {
                case UInt8(ascii: "\\"):
                    i += 1
                    if i < srclen && src[i] == UInt8(ascii: "\"") {
                        out.append(src[i])
                        i += 1
                    } else {
                        out.append(UInt8(ascii: "\\"))
                    }
                case UInt8(ascii: "\""):
                    i += 1
                    reading = false
                default:
                    out.append(src[i])
                    i += 1
                }
            }

        default:
            // Unquoted value
            var reading = true
            while i < srclen && src[i] != IAC && reading {
                switch src[i] {
                case UInt8(ascii: "}"), UInt8(ascii: "]"), UInt8(ascii: ","), UInt8(ascii: ":"):
                    reading = false
                case UInt8(ascii: " "):
                    i += 1
                default:
                    out.append(src[i])
                    i += 1
                }
            }
        }
    }

    out.append(IAC)
    out.append(SE)
    return out
}

File: kotlin/mth-core/src/main/kotlin/mth/core/MSDPConverter.kt
package mth.core

// MSDP protocol control bytes
private const val MSDP_VAR: Byte = 1
private const val MSDP_VAL: Byte = 2
private const val MSDP_TABLE_OPEN: Byte = 3
private const val MSDP_TABLE_CLOSE: Byte = 4
private const val MSDP_ARRAY_OPEN: Byte = 5
private const val MSDP_ARRAY_CLOSE: Byte = 6

// Telnet framing bytes
private const val IAC: Byte = 0xFF.toByte()  // -1 signed
private const val SB: Byte = 0xFA.toByte()   // -6 signed
private const val SE: Byte = 0xF0.toByte()   // -16 signed
private const val TELOPT_MSDP: Byte = 69
private const val TELOPT_GMCP: Byte = 0xC9.toByte() // 201 / -55 signed

/** Extension to get unsigned int value of a Byte. */
private val Byte.u: Int get() = toInt() and 0xFF

/**
 * Convert MSDP binary subnegotiation to GMCP JSON subnegotiation.
 *
 * Input: `IAC SB TELOPT_MSDP <msdp data> IAC SE`
 * Output: `IAC SB TELOPT_GMCP MSDP {<json>} IAC SE`
 */
fun msdp2json(src: ByteArray): ByteArray {
    val out = mutableListOf<Byte>()
    val srclen = src.size

    // If MSDP framing, replace with GMCP framing + "MSDP {"
    if (srclen >= 3 && src[2] == TELOPT_MSDP) {
        out.add(IAC)
        out.add(SB)
        out.add(TELOPT_GMCP)
        for (b in "MSDP {".toByteArray(Charsets.UTF_8)) {
            out.add(b)
        }
    }

    var i = 3
    var nest = 0
    var last: Byte = 0

    while (i < srclen) {
        if (src[i] == IAC && i + 1 < srclen && src[i + 1] == SE) {
            break
        }

        when (src[i]) {
            MSDP_TABLE_OPEN -> {
                out.add('{'.code.toByte())
                nest++
                last = MSDP_TABLE_OPEN
            }
            MSDP_TABLE_CLOSE -> {
                if (last == MSDP_VAL || last == MSDP_VAR) {
                    out.add('"'.code.toByte())
                }
                if (nest > 0) nest--
                out.add('}'.code.toByte())
                last = MSDP_TABLE_CLOSE
            }
            MSDP_ARRAY_OPEN -> {
                out.add('['.code.toByte())
                nest++
                last = MSDP_ARRAY_OPEN
            }
            MSDP_ARRAY_CLOSE -> {
                if (last == MSDP_VAL || last == MSDP_VAR) {
                    out.add('"'.code.toByte())
                }
                if (nest > 0) nest--
                out.add(']'.code.toByte())
                last = MSDP_ARRAY_CLOSE
            }
            MSDP_VAR -> {
                if (last == MSDP_VAL || last == MSDP_VAR) {
                    out.add('"'.code.toByte())
                }
                if (last == MSDP_VAL || last == MSDP_VAR || last == MSDP_TABLE_CLOSE || last == MSDP_ARRAY_CLOSE) {
                    out.add(','.code.toByte())
                }
                out.add('"'.code.toByte())
                last = MSDP_VAR
            }
            MSDP_VAL -> {
                if (last == MSDP_VAR) {
                    out.add('"'.code.toByte())
                    out.add(':'.code.toByte())
                }
                if (last == MSDP_VAL) {
                    out.add('"'.code.toByte())
                    out.add(','.code.toByte())
                }
                if (i + 1 < srclen && src[i + 1] != MSDP_TABLE_OPEN && src[i + 1] != MSDP_ARRAY_OPEN) {
                    out.add('"'.code.toByte())
                }
                last = MSDP_VAL
            }
            '\\'.code.toByte() -> {
                out.add('\\'.code.toByte())
                out.add('\\'.code.toByte())
            }
            '"'.code.toByte() -> {
                out.add('\\'.code.toByte())
                out.add('"'.code.toByte())
            }
            else -> {
                out.add(src[i])
            }
        }
        i++
    }

    // Append closing "}" and IAC SE
    out.add('}'.code.toByte())
    out.add(IAC)
    out.add(SE)

    return out.toByteArray()
}

/**
 * Convert GMCP JSON subnegotiation to MSDP binary subnegotiation.
 *
 * Input: `IAC SB TELOPT_GMCP MSDP {<json>} IAC SE`
 * Output: `IAC SB TELOPT_MSDP <msdp data> IAC SE`
 */
fun json2msdp(src: ByteArray): ByteArray {
    val out = mutableListOf<Byte>()
    val srclen = src.size

    // If GMCP framing, replace with MSDP framing
    if (srclen >= 3 && src[2] == TELOPT_GMCP) {
        out.add(IAC)
        out.add(SB)
        out.add(TELOPT_MSDP)
    }

    var i = 3

    // Skip "MSDP {" prefix if present
    if (i + 6 <= srclen) {
        val prefix = "MSDP {".toByteArray(Charsets.UTF_8)
        var matches = true
        for (k in prefix.indices) {
            if (src[i + k] != prefix[k]) {
                matches = false
                break
            }
        }
        if (matches) {
            i += 6
        }
    }

    val state = IntArray(100)
    var nest = 0
    var last: Byte = 0
    state[0] = 0

    while (i < srclen && src[i] != IAC && nest < 99) {
        when (src[i]) {
            ' '.code.toByte() -> {
                i++
            }
            '{'.code.toByte() -> {
                out.add(MSDP_TABLE_OPEN)
                i++
                nest++
                state[nest] = 0
            }
            '}'.code.toByte() -> {
                nest--
                i++
                if (nest < 0) {
                    out.add(IAC)
                    out.add(SE)
                    return out.toByteArray()
                }
                out.add(MSDP_TABLE_CLOSE)
            }
            '['.code.toByte() -> {
                i++
                nest++
                state[nest] = 1
                out.add(MSDP_ARRAY_OPEN)
            }
            ']'.code.toByte() -> {
                nest--
                i++
                out.add(MSDP_ARRAY_CLOSE)
            }
            ':'.code.toByte() -> {
                out.add(MSDP_VAL)
                i++
            }
            ','.code.toByte() -> {
                i++
                if (state[nest] != 0) {
                    out.add(MSDP_VAL)
                } else {
                    out.add(MSDP_VAR)
                }
            }
            '"'.code.toByte() -> {
                i++
                if (last == 0.toByte()) {
                    last = MSDP_VAR
                    out.add(MSDP_VAR)
                }

                // Read quoted string
                var reading = true
                while (i < srclen && src[i] != IAC && reading) {
                    when (src[i]) {
                        '\\'.code.toByte() -> {
                            i++
                            if (i < srclen && src[i] == '"'.code.toByte()) {
                                out.add(src[i])
                                i++
                            } else {
                                out.add('\\'.code.toByte())
                            }
                        }
                        '"'.code.toByte() -> {
                            i++
                            reading = false
                        }
                        else -> {
                            out.add(src[i])
                            i++
                        }
                    }
                }
            }
            else -> {
                // Unquoted value
                var reading = true
                while (i < srclen && src[i] != IAC && reading) {
                    when (src[i]) {
                        '}'.code.toByte(), ']'.code.toByte(), ','.code.toByte(), ':'.code.toByte() -> {
                            reading = false
                        }
                        ' '.code.toByte() -> {
                            i++
                        }
                        else -> {
                            out.add(src[i])
                            i++
                        }
                    }
                }
            }
        }
    }

    out.add(IAC)
    out.add(SE)
    return out.toByteArray()
}
